File: src/main/java/frc/utils/LEDAnim/LEDAnim.java
package frc.utils.LEDAnim;

import edu.wpi.first.wpilibj.util.Color;

public class LEDAnim {
    public Color[] leds;

    private double scroll = 0;
    private double scrollShift;

    public LEDAnim(int length){
        leds = new Color[length];
        for(int i=0; i<length; i++){
            leds[i] = Color.kBlack;
        }
    }
    public LEDAnim(int length, Color c){
        leds = new Color[length];
        for(int i=0; i<length; i++){
            leds[i] = c;
        }
    }
    public LEDAnim(Color[] c){
        leds = c;
    }

    public Color[] getLEDs(){
        return leds;
    }
    public void update(){
        if(scroll != 0){
            scrollShift += scroll * 0.02;
            leds = offset(this, (int)scrollShift, true).getLEDs();
        }
    }

    /**
     * scroll at reletive speed
     * @param speed leds per second
     */
    public LEDAnim scroll(double speed){
        scroll = speed;
        return this;
    }
    /**
     * multiplies rgb values of base and mask
     * @param base
     * @param mask
     */
    static LEDAnim mask(LEDAnim base, LEDAnim mask){
        Color[] result = new Color[base.leds.length];
        for(int i=0; i<base.leds.length; i++){
            double r = (base.leds[i].red * mask.leds[i].red);
            double g = (base.leds[i].green * mask.leds[i].green);
            double b = (base.leds[i].blue * mask.leds[i].blue);
            result[i] = new Color(r, g, b);
        }

        return new LEDAnim(result);
    }

    public static LEDAnim offset(LEDAnim in, int offset, boolean wrap){
        Color[] result = new Color[in.leds.length];
        for(int i=0; i<in.leds.length; i++){
            if(wrap){
                result[i] = in.leds[(i - (int)offset + in.leds.length) % in.leds.length];
            } else {
                if(offset < 0 || offset >= in.leds.length){
                    result[i] = Color.kBlack;
                } else {
                    result[i] = in.leds[i+offset];
                }
            }
        }
        return new LEDAnim(result);
    }
}

File: src/main/kotlin/frc/utils/LEDAnim/LEDAnim.kt
package frc.utils.LEDAnim

import edu.wpi.first.wpilibj.util.Color
import frc.robot.constants.Constants
import frc.utils.ExtraMath

open class LEDAnim {
    @JvmField var leds: Array<Color>
    @JvmField var initLeds: Array<Color>? = null

    private var scroll: Double = 0.0
    private var scrollShift: Double = 0.0

    constructor(length: Int) {
        leds = Array(length) { Color.kBlack }
    }

    constructor(length: Int, c: Color) {
        leds = Array(length) { c }
    }

    constructor(c: Array<Color>) {
        leds = c
        initLeds = c.clone()
    }

    fun getLEDs(): Array<Color> = leds

    fun update() {
        if (scroll != 0.0) {
            scrollShift += scroll * Constants.EVENT_LOOP_TIME
            leds = offset(LEDAnim(initLeds!!.clone()), scrollShift, true).getLEDs()
        }
    }

    /**
     * scroll at relative speed
     * @param speed leds per second
     */
    fun scroll(speed: Double): LEDAnim {
        scroll = speed
        return this
    }

    companion object {
        /**
         * multiplies rgb values of base and mask
         */
        @JvmStatic
        fun mask(base: LEDAnim, mask: LEDAnim): LEDAnim {
            val result = Array(base.leds.size) { i ->
                val r = base.leds[i].red * mask.leds[i].red
                val g = base.leds[i].green * mask.leds[i].green
                val b = base.leds[i].blue * mask.leds[i].blue
                Color(r, g, b)
            }
            return LEDAnim(result)
        }

        @JvmStatic
        fun offset(`in`: LEDAnim, offset: Double, wrap: Boolean): LEDAnim {
            val result = Array(`in`.leds.size) { i ->
                if (wrap) {
                    val a = (i + offset) % `in`.leds.size
                    ExtraMath.colLerp(
                        `in`.leds[a.toInt()],
                        `in`.leds[(a + 1).toInt() % `in`.leds.size],
                        a - a.toInt()
                    )
                } else {
                    if (offset < 0 || offset >= `in`.leds.size) {
                        Color.kBlack
                    } else {
                        ExtraMath.colLerp(
                            `in`.leds[(i - offset).toInt()],
                            `in`.leds[((i - offset + 1).toInt()) % `in`.leds.size],
                            (i - offset) - (i - offset).toInt()
                        )
                    }
                }
            }
            return LEDAnim(result)
        }
    }
}
